### PR TITLE
[BREAKING] Remove unidiomatic numerical types from Go SDK

### PR DESCRIPTION
**Pull request by EvanBoyle**
Half of our base types in the Go SDK are numerical variants:

-  uint, ..., uint64, int8, ..., int64, float32 (10/20 base types that we generate complex types from). 

This change removes all numerical output types except for those that we actually use in providers. We're just left with `int` and `float64`. (I verified this with some searches, eg: https://github.com/search?q=org%3Apulumi+%22pulumi.uint%22&type=code)

Doing this cuts the base size of resulting Go binaries from __59M to 31M__. A comment below from a user reports a reduction from __450M to 250M__ building against this change. 

This is technically a breaking change, but these types are not exactly idiomatic given that our fundamental resource type system doesn't support these. We'll instead leave the user to convert to/from types supported by the system.

This gets us a huge reduction in binary size which has been a point of pain. I also observed a decrease in compilation times locally. Hopefully, this will also result in lower memory requirements to build, which we have heard complaints of OOMs in CI (and experienced this ourselves in linting, etc). 

I verified with a handful of heavy Go SDK users that removing these types won't break them. We'll announce this in community slack after the change is merged. Luckily this change is pretty easy to reverse. If we end up running into some dissat related to this change, we can always revert it and hold off until 3.0. Build time/memory/size has been a big pain point reported by multiple users, and I expect that users will overwhelmingly be very happy to see this.

Internal discussions pointed towards biasing to take this breaking change now and deliver this value to our users rather than holding it back until 3.0.

### CODE REVIEW

**Comment by EvanBoyle** [comment]
Other interesting notes. Binary size seems to degrade exponentially. 
- base pulumi program with change: `31M`
- base pulumi program without change: `59M`
- add S3 to program with change `44M` (`+13M`)
- add s3 to program without change `83M` (`+24M`)

**Comment by EvanBoyle** [comment]
Prototyped the improvements for removing contextful methods in this branch, but they are far smaller in magnitude: 

https://github.com/pulumi/pulumi/pull/6144

**Comment by yarinm** [comment]
The binary size decrease is amazing, my binaries went down from ~450mb to ~250mb

File: sdk/go/pulumi/generate.go [hunk 1]
```diff
@@ -58,8 +58,7 @@ func (b builtin) DefineInputMethods() bool {
 
 func (b builtin) ImplementsPtrType() bool {
 	switch b.Type {
-	case "bool", "float32", "float64", "ID", "int", "int16", "int32", "int64", "int8", "string", "URN", "uint",
-		"uint16", "uint32", "uint64", "uint8":
+	case "bool", "float64", "ID", "int", "string", "URN":
```
Comment (lblackstone) [inline sdk/go/pulumi/generate.go:61]: Should we keep `uint` around as well? Seems like we might want at least one unsigned type.

**Review by lblackstone — APPROVED** [review]

File: sdk/go/pulumi/generate.go (hunk 1 above)
Comment (EvanBoyle) [reply sdk/go/pulumi/generate.go:61]: I'd bias to not add any types unless we have a concrete use case driving it. Remember that these base types generate an explosion of map types, array types, pointer types, etc as well. Fairly easy to add back later if need be.

Comment (lblackstone) [reply sdk/go/pulumi/generate.go:61]: Fair enough. LGTM

File: sdk/go/pulumi/generate.go [hunk 2]
```diff
@@ -120,22 +119,12 @@ var builtins = makeBuiltins([]*builtin{
 	{Name: "Asset", Type: "Asset", inputType: "*asset", implements: []string{"AssetOrArchive"}, Example: "NewFileAsset(\"foo.txt\")"},
 	{Name: "AssetOrArchive", Type: "AssetOrArchive", Example: "NewFileArchive(\"foo.zip\")"},
 	{Name: "Bool", Type: "bool", Example: "Bool(true)", GenerateConfig: true, DefaultConfig: "false"},
-	{Name: "Float32", Type: "float32", Example: "Float32(1.3)", GenerateConfig: true, DefaultConfig: "0"},
 	{Name: "Float64", Type: "float64", Example: "Float64(999.9)", GenerateConfig: true, DefaultConfig: "0"},
 	{Name: "ID", Type: "ID", inputType: "ID", implements: []string{"String"}, Example: "ID(\"foo\")"},
```
Comment (pgavlin) [inline sdk/go/pulumi/generate.go:123]: We should consider removing the `ID` and `URN` types from the SDK as well, though this might be a larger break.

**Review by pgavlin — APPROVED** [review]
LGTM

File: sdk/go/pulumi/generate.go (hunk 2 above)
Comment (EvanBoyle) [reply sdk/go/pulumi/generate.go:123]: Just opt for plain strings instead for the purposes of perf? Probably results in a usability improvement as well.

**Pull request merged**